cmd/validate.go

```diff
@@ -0,0 +1,272 @@
+	if err := registry.RegisterPlugin(root, serverOpts.PluginConfig[plug]); err != nil {
+		cmdutil.ErrPrintf("Error loading plugin: %v\n", err)
+		return exitCode{1}
+	}
+
+	rand.Seed(time.Now().UnixNano())
+	plugin.InitCache()
+	ctx := context.Background()
+
+	pw := progress.NewWriter()
+	pw.SetUpdateFrequency(50 * time.Millisecond)
+	pw.Style().Colors = progress.StyleColorsExample
+	go pw.Render()
+
+	// Use list to walk the hierarchy breadth-first. If a schema is present, visit each unique type
+	// at each level once, and stop when we've fully explored the schema. If there are errors, print
+	// all the detail we have about them. Use an unbuffered channel to register errors as they come in.
+	erred := 0
+	errs := make(chan error)
+	go func() {
+		for err := range errs {
+			erred++
+			cmdutil.ErrPrintf("%v\n", err)
+		}
+	}()
+
+	// We use a worker pool to limit work-in-progress, and a wait group to know when all queued work
+	// is complete. Put the plugin on the worker pool. Listing the registry should never error.
+	wp := newPool(parallel)
+	entries, _ := plugin.CachedList(ctx, registry)
+	for _, e := range entries {
+		wp.Submit(func() { processEntry(ctx, pw, wp, e, all, errs) })
+	}
+
+	// Wait for the workgroup's we've queued to finish, then stop the worker pool.
+	wp.Finish()
+	// Leave time for progress to finish rendering.
+	time.Sleep(100 * time.Millisecond)
+	pw.Stop()
+	if erred > 0 {
+		cmdutil.ErrPrintf("Found %v errors.\n", erred)
+		return exitCode{1}
+	}
+
+	cmdutil.Println("Looks good!")
+	return exitCode{0}
+}
+
+type pool struct {
+	wp *workerpool.WorkerPool
+	wg *sync.WaitGroup
+}
+
+func newPool(parallel int) pool {
+	return pool{wp: workerpool.New(parallel), wg: &sync.WaitGroup{}}
+}
+
+func (p pool) Submit(f func()) {
+	p.wg.Add(1)
+	p.wp.Submit(f)
+}
+
+func (p pool) Done() {
+	p.wg.Done()
+}
+
+func (p pool) Finish() {
+	// Wait for the workgroup's we've queued to finish, then stop the worker pool.
+	p.wg.Wait()
+	p.wp.StopWait()
+}
+
+type criteria struct {
+	list, read, stream, exec bool
+}
+
+func newCriteria(entry plugin.Entry) criteria {
+	var crit criteria
+	crit.list = plugin.ListAction().IsSupportedOn(entry)
+	crit.read = plugin.ReadAction().IsSupportedOn(entry)
+	crit.stream = plugin.StreamAction().IsSupportedOn(entry)
+	crit.exec = plugin.ExecAction().IsSupportedOn(entry)
+	return crit
+}
+
+func (c criteria) String() string {
+	s := []byte("----")
+	if c.list {
+		s[0] = 'l'
+	}
+
+	if c.read {
+		s[1] = 'r'
+	}
+
+	if c.stream {
+		s[2] = 's'
+	}
+
+	if c.exec {
+		s[3] = 'x'
+	}
+	return string(s)
+}
+
+func processEntry(ctx context.Context, pw progress.Writer, wp pool, e plugin.Entry, all bool, errs chan<- error) {
+	defer wp.Done()
+	name := plugin.ID(e)
+	crit := newCriteria(e)
+	tracker := progress.Tracker{Message: fmt.Sprintf("Testing %s %s", crit, name), Total: 4}
+	pw.AppendTracker(&tracker)
+
+	if plugin.ListAction().IsSupportedOn(e) {
+		entries, err := plugin.CachedList(ctx, e.(plugin.Parent))
```

Comment:
Yeah, I guess starting a whole server up would be overkill. At the same time, breaking consistency with the other commands + seeing API-specific stuff like ErrorObj move over to plugin is also kind of weird. Not using a server also means we'll need to handle context-cancellation related cleanup (like on Ctrl-C). I don't know how hard the latter would be though.

How horrible would a "light-weight" server (just the API instead of FUSE) be?

Reply:
"API-specific" stuff like ErrorObj was moved over so I had a way to return structured information from externalPluginEntry, not anything related to using a server. I also don't see `validate` as something you should use from within the wash shell. I started with an API server, but it seemed like overkill.

I'll add the context cancellation.

Reply:
Ok, that's fair.